proposals/p5434.md

```diff
@@ -0,0 +1,1331 @@
+-   for forwarding arguments while preserving
+    [expression category](/docs/design/README.md#expression-categories);
+-   to add a feature to pattern matching to modify things after they have been
+    matched;
+-   to support refactoring code without changing all the uses of a name, a
+    problem we are already seeing with `self` and `addr self`, and would be a
+    point of friction in local pattern matching in the future; and
+-   to support breaking up an expression into pieces without altering the
+    expression category of individual pieces.
+
+Reference returns have also come up before, particularly to support operators
+such as indexing `[`...`]` and other functions that should produce a reference
+expression. It is desirable, though, that this not introduce new memory unsafety
+concerns, due to returning a reference to something with insufficient lifetime.
+
+In addition, we have been interested in adding other return mechanisms that
+support returning values in registers in cases that our current convention
+won't.
+
+## Background
+
+-   Carbon has
+    [reference expressions](/docs/design/values.md#reference-expressions).
+-   Using
+    [the `addr` keyword on mutating methods to get a `self` with a pointer type](/docs/design/classes.md#methods)
+    was introduced in
+    [proposal #722: "Nominal classes and methods"](/proposals/p0722.md#keyword-to-indicate-pass-by-address).
+-   [Leads issue #5261: "We should add `ref` bindings to Carbon, paralleling reference expressions"](https://github.com/carbon-language/carbon-lang/issues/5261)
+    supports adding `ref` bindings to Carbon.
+-   [LLVM's `noalias` attribute](https://llvm.org/docs/LangRef.html#function-attributes)
+    is used to mark a pointer as being aliased in only limited ways to enable
+    optimization. Also see
+    [LLVM's pointer aliasing rules](https://llvm.org/docs/LangRef.html#pointer-aliasing-rules).
+-   Marking a pointer as not captured, to allow optimizations, was originally
+    done with
+    [LLVM's `nocapture` attribute](https://releases.llvm.org/11.0.0/docs/LangRef.html#parameter-attributes),
+    which has become
+    [`captures(none)` and `captures(ret: address, provenance)`](https://llvm.org/docs/LangRef.html#function-attributes),
+    which is governed by
+    [pointer capture rules](https://llvm.org/docs/LangRef.html#pointer-capture).
+-   Clang allows C++ code to use the
+    [`clang::lifetimebound` attribute](https://clang.llvm.org/docs/AttributeReference.html#id8)
+    to mark parameters that may be referenced by the return value, in order to
+    detect some classes of use-after-free memory-safety bugs.
+-   [C++ has reference types](https://en.cppreference.com/w/cpp/language/reference).
+
+## Proposal
+
+### `ref` bindings
+
+We introduce a new keyword `ref`. This may be added to a `:` binding to mark it
+as binding to a reference expression, as in:
+
+```carbon
+fn F(ptr: i32*) {
+  // A reference binding `x`.
+  let ref x: i32 = *ptr;
```

Comment:
At a high level, `let` can be thought of as a const thing, and `var` as a mutable thing, though there are of course storage implications. But this breaks that mental model for developers, so maybe we could preserve that by using `ref x:` instead?

Reply:
I think this has been resolved with the move to `-> val T`, and marking resolved. Re-open if I'm wrong though!